resizing_hashtable/r_hashtables.py

```diff
@@ -9,56 +9,102 @@ def __init__(self, key, value):
         self.value = value
         self.next = None
 
-
 # '''
 # Fill this in
 
 # Resizing hash table
 # '''
 class HashTable:
     def __init__(self, capacity):
-        pass
-
+        self.capacity = capacity
+        self.storage = [None] * self.capacity
 
 # '''
 # Research and implement the djb2 hash function
 # '''
 def hash(string, max):
-    pass
-
+    hashNum = 5381
+    for x in string:
+        hashNum = (( hashNum << 5) + hashNum) + ord(x)
+    return hashNum % max
 
 # '''
 # Fill this in.
 
 # Hint: Used the LL to handle collisions
 # '''
 def hash_table_insert(hash_table, key, value):
-    pass
-
+    index = hash(key, hash_table.capacity)
+    pair = LinkedPair(key, value)
+    if hash_table.storage[index] is None:
+        hash_table.storage[index] = pair
+        return
+    current = hash_table.storage[index]
+    while current.next is not None:
+        if current.key != key:
+            current = current.next
+        else:
+            current.value = value
+            return
+    if current.key == key:
+        current.value = value
+        return
+    current.next = pair
+    return
 
 # '''
 # Fill this in.
 
 # If you try to remove a value that isn't there, print a warning.
 # '''
 def hash_table_remove(hash_table, key):
-    pass
-
+    index = hash(key, hash_table.capacity)
+    if hash_table.storage[index] is None:
+        print("Error in Remove")
+    current = hash_table.storage[index]
+    prev = None
+    while current is not None:
+        if current.key == key:
+            if prev:
+                prev.next = current.next
+                current.next = None
+            else:
+                hash_table.storage[index] = None
+        prev = current
+        current = current.next
+    
+            
 
 # '''
 # Fill this in.
 
 # Should return None if the key is not found.
 # '''
 def hash_table_retrieve(hash_table, key):
-    pass
+    index = hash(key, hash_table.capacity)
+    if hash_table.storage[index] is None:
+        return None
+    current = hash_table.storage[index]
+    while current is not None:
+        if current.key == key:
+            return current.value
+        current = current.next
+    return None
 
 
 # '''
 # Fill this in
 # '''
 def hash_table_resize(hash_table):
-    pass
+    new_capacity = hash_table.capacity * 2
+    new_storage = [None] * new_capacity
+
+    # copy over elements
+    for i in range(0, len(hash_table.storage)):
+        new_storage[i] = hash_table.storage[i]
```

Comment:
Re-iterating what we discussed in 1-1:

This resize function will need to rehash each node by its key and insert it into the new storage system according to the index that comes from that hash.  The index's that each node is inserted to will be different, and the linked lists that are created from collision will be different as well.  Currently we are just copying the linked lists from one array to the other at the same index.